src/openms/source/METADATA/SpectrumLookup.cpp

```diff
@@ -0,0 +1,276 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Hendrik Weisser $
+// $Authors: Hendrik Weisser $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/METADATA/SpectrumLookup.h>
+
+using namespace std;
+
+namespace OpenMS
+{
+  const String& SpectrumLookup::default_scan_regexp = "=(?<SCAN>\\d+)$";
+
+  const String& SpectrumLookup::regexp_names_ = "INDEX0 INDEX1 SCAN ID RT";
+
+  SpectrumLookup::SpectrumLookup(): 
+    rt_tolerance(0.01), n_spectra_(0),
+    regexp_name_list_(ListUtils::create<String>(regexp_names_, ' '))
+  {}
+
+
+  SpectrumLookup::~SpectrumLookup()
+  {}
+
+
+  bool SpectrumLookup::empty() const
+  {
+    return n_spectra_ == 0;
+  }
+  
+
+  Size SpectrumLookup::findByRT(double rt) const
+  {
+    double upper_diff = numeric_limits<double>::infinity();
+    map<double, Size>::const_iterator upper = rts_.upper_bound(rt);
+    if (upper != rts_.end())
+    {
+      upper_diff = upper->first - rt;
+    }
+    double lower_diff = numeric_limits<double>::infinity();
+    map<double, Size>::const_iterator lower = upper;
+    if (lower != rts_.begin())
+    {
+      --lower;
+      lower_diff = rt - lower->first;
+    }
+    if ((lower_diff < upper_diff) && (lower_diff <= rt_tolerance))
+    {
+      return lower->second;
+    }
+    if (upper_diff <= rt_tolerance) return upper->second;
+
+    String element = "spectrum with RT " + String(rt);
+    throw Exception::ElementNotFound(__FILE__, __LINE__, __PRETTY_FUNCTION__,
+                                     element);
+  }
+
+
+  Size SpectrumLookup::findByNativeID(const String& native_id) const
+  {
+    map<String, Size>::const_iterator pos = ids_.find(native_id);
+    if (pos == ids_.end())
+    {
+      String element = "spectrum with native ID '" + native_id + "'";
+      throw Exception::ElementNotFound(__FILE__, __LINE__, __PRETTY_FUNCTION__,
+                                       element);
+    }
+    return pos->second;
+  }
+
+
+  Size SpectrumLookup::findByIndex(Size index, bool count_from_one) const
+  {
+    Size adjusted_index = index;
+    if (count_from_one) --adjusted_index; // overflow (index = 0) handled below
```

Comment:
Next line. ;-)
If it overflows ("underflows"?), `adjusted_index` will be the max. `Size` value, so always greater than `n_spectra_`.